planning/MartenV4.md

```diff
@@ -0,0 +1,454 @@
+
+1. The inline projections work across the incoming event objects, create new projected documents
+or update existing documents as necessary just prior to sending the current `UnitOfWork` to the database
+1. The `DocumentSession` issues the batched up commands with both the event capture and the projection documents
+1. The underlying database transaction is committed
+
+If there's no need for event metadata, do thing in fewer network round trips
+
+#### If event or stream metadata is necessary...
+
+This gets a little more complicated, because the projections can't be calculated until *after* the events are appended, so:
+
+1. The events get appended, and the sproc returns an array of the new event versions
+1. Build up an array of `IEvent` wrappers that have the stream and event metadata
+1. Pass the `IEvent` wrappers into the transformation code, and get an additional `UpdateBatch` of projected document inserts
+   and updates
+1. Submit the subsequent `UpdateBatch` to the database
+1. The underlying database transaction is committed
+
+The difference here is that this will take multiple batched commands to the database within the same `NpgsqlTransaction`
+
+
+
+
+### Snapshots & "Live" Projections
+
+TODO
+
+## Async Daemon
+
+For v4, the Async Daemon needs to come in a couple different possible deployment configurations:
+
+1. "Classic" Polling-Based Mode -- no queueing, just something like the existing Async Daemon that polls for new events and runs the projections. I think this will be a near rewrite in v4, but there's some 
+   existing art in [Jasper's DurabilityAgent](http://jasperfx.github.io/documentation/durability/) that might help this along
+1. Messaging-Based Mode -- use Rabbit MQ, Azure Service Bus, Kafka, etc. to gather up events and connect with the running projections
+
+For both modes, we've got these issues to solve:
+
+1. Assigning projection ownership between active nodes, including the distribution and segmentation of projection builders
+1. Triggering a projection rebuild? Or at least a switchover?
+1. Enforcing event ordering
+
+Some general thoughts:
+
+* Try to decouple the projections away from `IDocumentSession` as a way to make the projection code more efficient
+* By skipping the document session, we could go directly at individual `IStorageOperation` for updates maybe. Allowing us to more effectively parallelize the generation of 
+  document updates
+* Generate a new Postgresql function for projected documents that combines the progress tracking by projection, adds optimistic concurrency checks,
+  and upserts an array of documents. Postgresql support for arrays makes this feasible
+* We need to support zero downtime projection rebuilds
+* For aggregations, introduce some caching on the aggregated documents as an optimization. Depending on the aggregation, this might not be super effective,
+  but would be huge for rebuilding projections
+
```

Comment:
I think that we should also think and deliver a concept of the pluggable chain of Upcasters and Downcasters (so as decorators for aggregation/projection). Currently to achieve that it's possible only via the JSON.NET magic. It would be not only useful to easier maintain the event versioning (so transforming event from the V2 version to V3). 
It could be also used for event encryption (eg. for GDPR).

Most of the Event Stores provide such functionality out of the box.

See also: https://github.com/JasperFx/marten/issues/1377.

Reply:
@oskardudycz In regards to encryption, I'm thinking if [pgcrypto](https://www.postgresql.org/docs/current/pgcrypto.html) could be utilized. So it would be somewhat transparently usable through Marten. Though it maybe should be raised as a separate talking point altogether.

For the other points, the zero downtime projection rebuilds to me feel like a huge usability improvement and a good selling point. Would reduce or even eliminate some dance with upgrades & maintenance.

Reply:
Using pgcrypto seems to be interesting idea. We can investigate that to have built in encryption mechanism.
However, having the decorators for the other cases is still a must have in my opinion.